### DISCUSSION TITLE
How to allocate JSON object on heap space?

### DISCUSSION THREAD

**Discussion by himanshusahu06** [danielaparker/jsoncons#446] [Q&A]
How do I allocate json in stack space? Current implementation `json j = json::parse(raw_string)` creates object in stack space which gets destroyed as soon as it leaves the context.

We can't use shared_ptr because underlying library doesn't support it.

I tried `json *j = new json()` but not sure how do I parse a json string and store it in `j`?

Below code works, but again it this will first create object on stack and then copy it to the j which is inefficient.

```
jsoncons::json *j = new jsoncons::json();
*j = jsoncons::json::parse(raw_string);
```

**Comment by himanshusahu06** [comment]
Hi @danielaparker!  Could you please help answering this question? Thanks!

**Comment by jsonconsorg** [comment]
Apologies, just got back from a long trek in the Pyrenees. You can do it as follows (without a copy)

```
#include <jsoncons/json.hpp>

int main()
{
    jsoncons::json doc = jsoncons::json::parse(R"({"foo" : "bar"})");

    jsoncons::json* pdoc = new jsoncons::json(std::move(doc));
}
```

**Reply by himanshusahu06** [reply]
Thanks @jsonconsorg! I hope you enjoyed the trek!

This worked but I'm curious to know that still calls destructor of `doc`. Does that mean it's not O(1) operation?

**Reply by jsonconsorg** [reply]
In the move constructor, `pdoc` acquires the resources previously held by `doc`, and `doc` becomes a JSON `null` value. So when the destructor of `doc` gets called, it has nothing to do, no resources to release.

**Comment by himanshusahu06** [comment]
Thanks! It helped!
